src/Lachain.Core/Blockchain/Operations/BlockManager.cs

```diff
@@ -150,7 +161,28 @@ public Block LatestBlock()
         [MethodImpl(MethodImplOptions.Synchronized)]
         public Block? GetByHeight(ulong blockHeight)
         {
-            return _stateManager.LastApprovedSnapshot.Blocks.GetBlockByHeight(blockHeight);
+            try
+            {
+                /* try to block from cache */
+                if (_heightCache.TryGetValue(blockHeight, out var block))
+                    return block;
+
+                var newBlock = _stateManager.LastApprovedSnapshot.Blocks.GetBlockByHeight(blockHeight);
+
+                /* if it reach cache size limit, remove the oldest one */
+                if (_heightCache.Count == _blockSizeLimit)
+                    _heightCache.Remove(_heightCache.First());
```

Comment:
This is actually O(# of elements). So, it would be very slow and not scalable if we want to store let's say 1 million items in cache. What do you think about it ?

Reply:
so block size limit is set as 100 by default.

Reply:
I mean - we should make the erase faster than O(# of elements). Can we use a sortedSet (https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.sortedset-1?view=net-6.0) instead of a dictionary ?

Reply:
Looks like Dictionary.Remove has O(1) efficiency,  https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.remove?view=net-5.0 : This method approaches an O(1) operation.

Reply:
Actually the culprint the first(). It's a linq expression which is O(n). Dictionary itself has no first() method.

Reply:
Hi @HMAshiqulIslam how about using ElementAt(0)?

Reply:
A dictionary is unordered. I don't think there is any way to find the minimum key without iterating over all the keys. 
So elementAt(0) should also work in O(n).

Reply:
@HMAshiqulIslam, @sgladkov, Dictionary in principle has a faster lookup with O(1) while the lookup performance of a List is an O(n) operation.
if we have to use SortedSet, we have to use ValueKeyPair, so for our case, SortedDictonary is better. but there is no way to get the element at 0, I will go with elementAt(0) or First().. looks same for me.
Please review my recent commit.

Reply:
@kazunetakeda25 
I think we still have the same issue. _heightCache.First().Key
What does First() really return ? How fast is this operation ?

Reply:
@HMAshiqulIslam Please share me a better option if you have any... I have no idea...

Reply:
I think the removal of the first element is always O(n).
Maybe we should think different approach... Please let me know @sgladkov @HMAshiqulIslam

Reply:
Are queue kinds helpful then?

Reply:
I am bit confused if for a sortedDictionary, .First() is O(n) or O(logN). 
One other approach that we can take is as follows: 
we keep the keys in a sortedSet and (key, value) to another dictionary. 
Then, we can do sortedSet.Min to find the minimum key and erase them from both the dictionary and the sortedSet.

Reply:
Also, for this scenerio, queues should also work. We can store the keys in a queue and a separate dictionary for (key, value) pairs and erase the key from the front of the queue. 
What do you think ? @kazunetakeda25

Reply:
Looks good to me @HMAshiqulIslam

Reply:
Hi @sgladkov @HMAshiqulIslam Please review my recent new commit.
I used Queue and all methods i used are O(1)

Reply:
This looks good to me. One thing to note : do you think we should implement an isolated generic FIFO cache and make use of this in the blockManager ? We will be using this cache a lot later in different methods.

Reply:
We can move buffer to separate abstraction in the next PR where we will use it for other method. Let's implement it in several iterations.

Reply:
I am currently working on it already. @sgladkov @HMAshiqulIslam